File: gitdot-server/src/handler/review.rs
mod add_review_reviewer;
mod approve_review_diff;
mod create_review_comments;
mod reject_review_diff;
mod get_review;
mod get_review_diff;
mod list_reviews;
mod merge_review;
mod merge_review_diff;
mod publish_review;
mod publish_review_diff;
mod remove_review_reviewer;
mod resolve_review_comment;
mod update_review;
mod update_review_comment;
mod update_review_diff;

use crate::app::AppState;
use axum::{
    Router,
    routing::{delete, get, patch, post},
};

use add_review_reviewer::add_review_reviewer;
use approve_review_diff::approve_review_diff;
use create_review_comments::create_review_comments;
use reject_review_diff::reject_review_diff;
use get_review::get_review;
use get_review_diff::get_review_diff;
use list_reviews::list_reviews;
use merge_review::merge_review;
use merge_review_diff::merge_review_diff;
use publish_review::publish_review;
use publish_review_diff::publish_review_diff;
use remove_review_reviewer::remove_review_reviewer;
use resolve_review_comment::resolve_review_comment;
use update_review::update_review;
use update_review_comment::update_review_comment;
use update_review_diff::update_review_diff;

pub fn create_review_router() -> Router<AppState> {
    Router::new()
        .route(
            "/repository/{owner}/{repo}/review/{number}",
            get(get_review).patch(update_review),
        )
        .route("/repository/{owner}/{repo}/reviews", get(list_reviews))
        .route(
            "/repository/{owner}/{repo}/review/{number}/publish",
            post(publish_review),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/reviewer",
            post(add_review_reviewer),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/reviewer/{reviewer_name}",
            delete(remove_review_reviewer),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/diff/{position}",
            get(get_review_diff).patch(update_review_diff),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/merge",
            post(merge_review),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/diff/{position}/merge",
            post(merge_review_diff),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/diff/{position}/publish",
            post(publish_review_diff),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/diff/{position}/approve",
            post(approve_review_diff),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/diff/{position}/reject",
            post(reject_review_diff),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/diff/{position}/comments",
            post(create_review_comments),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/comment/{comment_id}",
            patch(update_review_comment),
        )
        .route(
            "/repository/{owner}/{repo}/review/{number}/comment/{comment_id}/resolve",
            post(resolve_review_comment),
        )
}

File: gitdot-web/app/dal/review.ts
import "server-only";

import type {
  JudgeReviewDiffRequest,
  PublishReviewRequest,
  UpdateReviewDiffRequest,
  UpdateReviewRequest,
} from "gitdot-api";
import {
  GetReviewDiffResponse,
  ReviewCommentResource,
  ReviewerResource,
  ReviewResource,
} from "gitdot-api";
import { z } from "zod";
import { toQueryString } from "@/util";
import {
  authDelete,
  authFetch,
  authPatch,
  authPost,
  GITDOT_SERVER_URL,
  handleEmptyResponse,
  handleResponse,
} from "./util";

export async function listReviews(
  owner: string,
  repo: string,
): Promise<ReviewResource[] | null> {
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/reviews`,
  );

  return await handleResponse(response, z.array(ReviewResource));
}

export async function getReview(
  owner: string,
  repo: string,
  number: number,
): Promise<ReviewResource | null> {
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}`,
  );

  return await handleResponse(response, ReviewResource);
}

export async function getReviewDiff(
  owner: string,
  repo: string,
  number: number,
  position: number,
  revision?: number,
  compareTo?: number,
): Promise<GetReviewDiffResponse | null> {
  const params: Record<string, number> = {};
  if (revision !== undefined) params.revision = revision;
  if (compareTo !== undefined) params.compare_to = compareTo;
  const query = toQueryString(params);
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/diff/${position}${query ? `?${query}` : ""}`,
  );

  return await handleResponse(response, GetReviewDiffResponse);
}

export async function addReviewer(
  owner: string,
  repo: string,
  number: number,
  request: { user_name: string },
): Promise<ReviewerResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/reviewer`,
    request,
  );

  return await handleResponse(response, ReviewerResource);
}

export async function removeReviewer(
  owner: string,
  repo: string,
  number: number,
  reviewerName: string,
): Promise<void> {
  const response = await authDelete(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/reviewer/${reviewerName}`,
  );

  await handleEmptyResponse(response);
}

export async function updateDiff(
  owner: string,
  repo: string,
  number: number,
  position: number,
  request: UpdateReviewDiffRequest,
): Promise<ReviewResource | null> {
  const response = await authPatch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/diff/${position}`,
    request,
  );

  return await handleResponse(response, ReviewResource);
}

export async function updateReview(
  owner: string,
  repo: string,
  number: number,
  request: UpdateReviewRequest,
): Promise<ReviewResource | null> {
  const response = await authPatch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}`,
    request,
  );

  return await handleResponse(response, ReviewResource);
}

export async function publishReview(
  owner: string,
  repo: string,
  number: number,
  request: PublishReviewRequest,
): Promise<ReviewResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/publish`,
    request,
  );

  return await handleResponse(response, ReviewResource);
}

export async function submitReview(
  owner: string,
  repo: string,
  number: number,
  position: number,
  request: JudgeReviewDiffRequest,
): Promise<ReviewResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/diff/${position}/submit`,
    request,
  );

  return await handleResponse(response, ReviewResource);
}

export async function mergeDiff(
  owner: string,
  repo: string,
  number: number,
  position: number,
): Promise<ReviewResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/diff/${position}/merge`,
    {},
  );

  return await handleResponse(response, ReviewResource);
}

export async function resolveReviewComment(
  owner: string,
  repo: string,
  number: number,
  commentId: string,
  resolved: boolean,
): Promise<ReviewCommentResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/review/${number}/comment/${commentId}/resolve`,
    { resolved },
  );

  return await handleResponse(response, ReviewCommentResource);
}
